source/common/grpc/transcoding_filter.cc

```diff
@@ -0,0 +1,396 @@
+    auto service = descriptor_pool_.FindServiceByName(service_name);
+    if (service == nullptr) {
+      throw EnvoyException("Could not find '" + service_name + "' in the proto descriptor");
+    }
+    for (int i = 0; i < service->method_count(); ++i) {
+      auto method = service->method(i);
+
+      auto http_rule = method->options().GetExtension(google::api::http);
+
+      log().debug("/" + service->full_name() + "/" + method->name());
+      log().debug(http_rule.DebugString());
+
+      switch (http_rule.pattern_case()) {
+      case ::google::api::HttpRule::kGet:
+        pmb.Register("GET", http_rule.get(), http_rule.body(), method);
+        break;
+      case ::google::api::HttpRule::kPut:
+        pmb.Register("PUT", http_rule.put(), http_rule.body(), method);
+        break;
+      case ::google::api::HttpRule::kPost:
+        pmb.Register("POST", http_rule.post(), http_rule.body(), method);
+        break;
+      case ::google::api::HttpRule::kDelete:
+        pmb.Register("DELETE", http_rule.delete_(), http_rule.body(), method);
+        break;
+      case ::google::api::HttpRule::kPatch:
+        pmb.Register("PATCH", http_rule.patch(), http_rule.body(), method);
+        break;
+      case ::google::api::HttpRule::kCustom:
+        pmb.Register(http_rule.custom().kind(), http_rule.custom().path(), http_rule.body(),
+                     method);
+        break;
+      default:
+        break;
+      }
+    }
+  }
+
+  path_matcher_ = pmb.Build();
+
+  type_helper_.reset(new google::grpc::transcoding::TypeHelper(
+      google::protobuf::util::NewTypeResolverForDescriptorPool(kTypeUrlPrefix, &descriptor_pool_)));
+
+  log().debug("transcoding filter loaded");
+}
+
+Status TranscodingConfig::CreateTranscoder(
+    const Http::HeaderMap& headers, ZeroCopyInputStream* request_input,
+    ::google::grpc::transcoding::TranscoderInputStream* response_input,
+    std::unique_ptr<Transcoder>& transcoder,
+    const google::protobuf::MethodDescriptor*& method_descriptor) {
+  std::string method = headers.Method()->value().c_str();
+  std::string path = headers.Path()->value().c_str();
+  std::string args;
+
+  size_t pos = path.find('?');
+  if (pos != std::string::npos) {
+    args = path.substr(pos + 1);
+    path = path.substr(0, pos);
+  }
+
+  RequestInfo request_info;
+  std::vector<VariableBinding> variable_bidings;
+  method_descriptor =
+      path_matcher_->Lookup(method, path, args, &variable_bidings, &request_info.body_field_path);
+  if (!method_descriptor) {
+    return Status(Code::NOT_FOUND, "Could not resolve " + path + " to a method");
+  }
+
+  auto status = MethodToRequestInfo(method_descriptor, &request_info);
+  if (!status.ok()) {
+    return status;
+  }
+
+  for (const auto& binding : variable_bidings) {
+    google::grpc::transcoding::RequestWeaver::BindingInfo resolved_binding;
+    auto status = type_helper_->ResolveFieldPath(*request_info.message_type, binding.field_path,
```

Comment:
Not a huge fan of variable shadowing (there is an outer status and inner status). Can we either use scopes to control the outer one of just have a single status variable that gets reused?